File: Lab5_maybeFinal/TestFileHandler.cpp
#include "TestFileHandler.h"
#include "FileHandler.h"
#include "Exceptions.h"
#include <iostream>
#include <fstream>
#include <stdexcept>
#include <exception>  // Для коректної обробки виключень

void TestFileHandler::test_load_addresses_valid_file() {
    try {
        std::ofstream file("addresses1.txt");
        if (!file.is_open()) {
            std::cerr << "Test failed: Unable to open file for writing.\n";
            return;
        }
        file << "Private: Kyiv Khreshchatyk 10\n";
        file << "Appartment: Lviv Shevchenka 5 12\n";
        file.close();

        FileHandler file_handler("addresses1.txt");
        auto addresses = file_handler.LoadAddresses();

        if (addresses.size() != 2) {
            std::cerr << "Test failed: Incorrect number of addresses loaded.\n";
            return;
        }

        if (addresses[0]->GetType() != "Private" || addresses[1]->GetType() != "Appartment") {
            std::cerr << "Test failed: Incorrect address types.\n";
            return;
        }

        std::cout << "Test passed: Valid file loading.\n";
    }
    catch (const std::exception& e) {
        std::cerr << "Test failed due to exception: " << e.what() << "\n";
    }
}

void TestFileHandler::test_load_addresses_file_not_found() {
    try {
        FileHandler file_handler("non_existent_file.txt");
        file_handler.LoadAddresses();
        std::cout << "Test failed: No exception thrown for non-existent file.\n";
    }
    catch (const NotFoundFileError&) {
        std::cout << "Test passed: Exception thrown for non-existent file.\n";
    }
    catch (const std::exception& e) {
        std::cerr << "Test failed due to exception: " << e.what() << "\n";
    }
}

void TestFileHandler::test_load_addresses_invalid_format() {
    try {
        std::ofstream file("addresses1.txt");
        if (!file.is_open()) {
            std::cerr << "Test failed: Unable to open file for writing.\n";
            return;
        }
        file << "Private: Kyiv Khreshchatyk 10\n";
        file << "InvalidLineWithoutColon\n";
        file.close();

        FileHandler file_handler("addresses1.txt");
        file_handler.LoadAddresses();
        std::cout << "Test failed: No exception thrown for invalid format.\n";
    }
    catch (const FileContentError&) {
        std::cout << "Test passed: Exception thrown for invalid format.\n";
    }
    catch (const std::exception& e) {
        std::cerr << "Test failed due to exception: " << e.what() << "\n";
    }
}

void TestFileHandler::test_load_addresses_invalid_number() {
    try {
        std::ofstream file("addresses1.txt");
        if (!file.is_open()) {
            std::cerr << "Test failed: Unable to open file for writing.\n";
            return;
        }
        file << "Private: Kyiv Khreshchatyk ABC\n";  // Некоректне число
        file.close();

        FileHandler file_handler("addresses1.txt");
        file_handler.LoadAddresses();
        std::cout << "Test failed: No exception thrown for invalid number.\n";
    }
    catch (const FileContentError&) {
        std::cout << "Test passed: Exception thrown for invalid number.\n";
    }
    catch (const std::exception& e) {
        std::cerr << "Test failed due to exception: " << e.what() << "\n";
    }
}

// Очищення файлу перед тестуванням
void clearFile(const std::string& filename) {
    std::ofstream file(filename, std::ios::trunc); // Очищаємо файл
    if (!file) {
        std::cerr << "Error clearing file: " << filename << std::endl;
    }
}

void TestFileHandler::test_save_address() {
    // Очищаємо файл перед тестом
    clearFile("addresses1.txt");
    try {
        PrivateHouse address("Kyiv", "Khreshchatyk", 10);
        FileHandler file_handler("addresses1.txt");

        file_handler.SaveAddress(address);

        std::ifstream file("addresses1.txt");
        std::string line;
        std::getline(file, line);
        if (line != "Private: Kyiv Khreshchatyk 10") {
            std::cout << "Test failed: Address not saved correctly.\n";
        }
        else {
            std::cout << "Test passed: Address saved correctly.\n";
        }
    }
    catch (const std::exception& e) {
        std::cerr << "Test failed due to exception: " << e.what() << "\n";
    }
}

void TestFileHandler::test_load_addresses_boundary_values() {
    try {
        // Тест для граничних значень номера будинку 
        std::ofstream file("addresses1.txt");
        if (!file.is_open()) {
            std::cerr << "Test failed: Unable to open file for writing.\n";
            return;
        }
        file << "Private: Kyiv Khreshchatyk " << std::numeric_limits<int>::min() << "\n"; // Мінімальне значення int (-2147483648)
        file << "Private: Lviv Shevchenka " << std::numeric_limits<int>::max() << "\n"; // Максимальне значення int (2147483647)
        file << "Private: Odesa Pushkina 0\n"; // Нульове значення
        file.close();

        FileHandler file_handler("addresses1.txt");
        auto addresses = file_handler.LoadAddresses();

        if (addresses.size() != 3) {
            std::cerr << "Test failed: Incorrect number of addresses loaded.\n";
            return;
        }

        // Перевірка, чи правильно завантажились дані
        if (addresses[0]->GetType() != "Private" || addresses[1]->GetType() != "Private" || addresses[2]->GetType() != "Private") {
            std::cerr << "Test failed: Incorrect address types.\n";
            return;
        }

        std::cout << "Test passed: Boundary values loaded correctly.\n";
    }
    catch (const std::exception& e) {
        std::cerr << "Test failed due to exception: " << e.what() << "\n";
    }
}

// Функція для запуску всіх тестів
void TestFileHandler::run_all_tests() {
    test_load_addresses_valid_file();
    test_load_addresses_file_not_found();
    test_load_addresses_invalid_format();
    test_load_addresses_invalid_number();
    test_save_address();
    test_load_addresses_boundary_values();
}

File: Lab5(maybe_final)/TestFileHandler.py
import unittest
from unittest.mock import mock_open, patch
from exceptions import NotFoundFileError, FileContentError
from address import Address
from appartment import Appartment
from private_house import PrivateHouse
from file_handler import FileHandler
import tempfile
import os

# C:\Users\User\Documents\Беляш\VS_code\Lab5
# python -m unittest TestFileHandler.py -v 

class TestFileHandler(unittest.TestCase):

    # Тест 1: Завантаження адрес із існуючого файлу
    def test_load_addresses_valid_file(self):
        # Тестові дані
        mock_file_content = "Private: Kyiv Khreshchatyk 10\nAppartment: Lviv Shevchenka 5 12\n"
        
        # Використовуємо mock_open для імітації відкриття файлу
        with patch("builtins.open", mock_open(read_data=mock_file_content)):
            file_handler = FileHandler(r"C:\Users\User\Documents\Беляш\VS_code\Lab5(maybe_final)\addresses.txt")  # використовуємо сировий рядок
            addresses = file_handler.load_addresses()
            
            self.assertEqual(len(addresses), 2)
            self.assertIsInstance(addresses[0], PrivateHouse)
            self.assertIsInstance(addresses[1], Appartment)

    # Тест 2: Завантаження адрес з файлу, якого не існує
    def test_load_addresses_file_not_found(self):
        with self.assertRaises(NotFoundFileError):
            file_handler = FileHandler(r"C:\Users\User\Documents\Беляш\VS_code\Lab5(maybe_final)\non_existent_file.txt")  # Вказуємо коректний шлях
            file_handler.load_addresses()

    # Тест 3: Завантаження файлу з некоректним форматом
    def test_load_addresses_invalid_format(self):
        mock_file_content = "Private: Kyiv Khreshchatyk 10\nInvalidLineWithoutColon\n"
        
        with patch("builtins.open", mock_open(read_data=mock_file_content)):
            file_handler = FileHandler(r"C:\Users\User\Documents\Беляш\VS_code\Lab5(maybe_final)\addresses.txt")
            with self.assertRaises(FileContentError):
                file_handler.load_addresses()

    # Тест 4: Завантаження файлу з некоректними числовими даними
    def test_load_addresses_invalid_number(self):
        mock_file_content = "Private: Kyiv Khreshchatyk ABC\n"
        
        with patch("builtins.open", mock_open(read_data=mock_file_content)):
            file_handler = FileHandler(r"C:\Users\User\Documents\Беляш\VS_code\Lab5(maybe_final)\addresses.txt")
            with self.assertRaises(FileContentError):
                file_handler.load_addresses()

    # Тест 5: Збереження адреси у файл
    def test_save_address(self):
        # Створення тестової адреси
        address = PrivateHouse("Kyiv", "Khreshchatyk", 10)

        # Використовуємо mock_open для перевірки запису у файл
        with patch("builtins.open", mock_open()) as mock_file:
            file_handler = FileHandler(r"C:\Users\User\Documents\Беляш\VS_code\Lab5(maybe_final)\addresses.txt")
            file_handler.save_address(address)
            
            # Перевіряємо, що файл відкривається з кодуванням 'windows-1251'
            mock_file.assert_called_once_with(r"C:\Users\User\Documents\Беляш\VS_code\Lab5(maybe_final)\addresses.txt", "a", encoding="utf-8")
            mock_file().write.assert_called_once_with(address.to_string() + '\n')
    
    # Тест 6: Граничні значення 
    def test_load_addresses_boundary_values(self):
        # Створення тимчасового файлу для тесту
        with tempfile.NamedTemporaryFile(delete=False, mode='w', encoding="utf-8") as temp_file:
            temp_filename = temp_file.name
            # Запис даних з граничними значеннями для тесту
            temp_file.write(f"Private: Kyiv Khreshchatyk {str(-2**31)}\n")  # Мінімальне значення int (-2147483648)
            temp_file.write(f"Private: Lviv Shevchenka {str(2**31-1)}\n")  # Максимальне значення int (2147483647)
            temp_file.write("Private: Odesa Pushkina 0\n")  # Нульове значення

        # Тепер завантажуємо адреси з цього тимчасового файлу
        file_handler = FileHandler(temp_filename)
        addresses = file_handler.load_addresses()

        # Перевірка кількості адрес
        self.assertEqual(len(addresses), 3)

        # Перевірка типу всіх адрес
        for address in addresses:
            self.assertIsInstance(address, PrivateHouse)

        # Перевірка значень адрес
        self.assertEqual(addresses[0].building_number, -2**31)
        self.assertEqual(addresses[1].building_number, 2**31-1)
        self.assertEqual(addresses[2].building_number, 0)

        # Видаляємо тимчасовий файл після тесту
        os.remove(temp_filename)

if __name__ == "__main__":
    unittest.main()
